src/stenotype/backend/grammar.py

```diff
@@ -0,0 +1,183 @@
+LIST = Group(Suppress("[") + TYPE + Suppress("]")).setParseAction(
+    lambda s, loc, toks: ste.List(toks[0][0])
+)
+DICT = Group(
+    Suppress("{") + TYPE + Suppress(":") + TYPE + Suppress("}")
+).setParseAction(lambda s, loc, toks: ste.Dict(toks[0][0], toks[0][1]))
+SET = Group(Suppress("{") + TYPE + Suppress("}")).setParseAction(
+    lambda s, loc, toks: ste.Set(toks[0][0])
+)
+
+CONTAINERS = TUPLE | LIST | DICT | SET
+
+# Literals
+# --------
+
+LITERAL_BOOL = (Keyword("True") | Keyword("False")).setParseAction(
+    lambda s, loc, toks: ste.Literal(True) if toks[0] == "True" else ste.Literal(False)
+)
+LITERAL_NONE = (Keyword("None")).setParseAction(lambda s, loc, toks: ste.Literal(None))
+LITERAL_ELLIPSIS = (Keyword("Ellipsis")).setParseAction(
+    lambda s, loc, toks: ste.Literal(Ellipsis)
+)
+LITERAL_INT = (Combine(Optional("-") + Word(nums))).setParseAction(
+    lambda s, loc, toks: ste.Literal(int(toks[0]))
+)
+_string_opts = dict(unquoteResults=False, convertWhitespaceEscapes=False)
+LITERAL_STR = Group(
+    QuotedString('"', escChar="\\", **_string_opts)
+    | QuotedString("'", escChar="\\", **_string_opts)
+).setParseAction(lambda s, loc, toks: ste.Literal(toks[0][0]))
+LITERAL_BYTES = Group(
+    "b"
+    + (
+        QuotedString('"', escChar="\\", **_string_opts)
+        | QuotedString("'", escChar="\\", **_string_opts)
+    )
+).setParseAction(lambda s, loc, toks: ste.Literal("b" + toks[0][1]))
+# TODO: float (does this even make sense? float is notoriously bad for precise values)
+
+KEYWORDS << (KEYWORDS.expr | LITERAL_BOOL | LITERAL_NONE | LITERAL_ELLIPSIS)
+LITERALS = (
+    LITERAL_BOOL
+    | LITERAL_NONE
+    | LITERAL_ELLIPSIS
+    | LITERAL_INT
+    | LITERAL_STR
+    | LITERAL_BYTES
+)
+
+# Shorthands
+# ----------
+
+ITERABLE = Group(Keyword("iter") + TYPE).setParseAction(
+    lambda s, loc, toks: ste.Iterable(toks[0][1])
+)
+CONTEXT = Group(Keyword("with") + TYPE).setParseAction(
+    lambda s, loc, toks: ste.Context(toks[0][1])
+)
+AWAITABLE = Group(Keyword("await") + TYPE).setParseAction(
+    lambda s, loc, toks: ste.Awaitable(toks[0][1])
+)
+ASYNC_ITERABLE = Group(Keyword("async iter") + TYPE).setParseAction(
+    lambda s, loc, toks: ste.AsyncIterable(toks[0][1])
+)
+ASYNC_CONTEXT = Group(Keyword("async with") + TYPE).setParseAction(
+    lambda s, loc, toks: ste.AsyncContext(toks[0][1])
+)
+
+KEYWORDS << (
+    KEYWORDS.expr
+    | ITERABLE.expr.exprs[0]
+    | CONTEXT.expr.exprs[0]
+    | AWAITABLE.expr.exprs[0]
+    | ASYNC_ITERABLE.expr.exprs[0]
+    | ASYNC_CONTEXT.expr.exprs[0]
+)
+SHORTHANDS = ITERABLE | CONTEXT | AWAITABLE | ASYNC_ITERABLE | ASYNC_CONTEXT
+
+#: any valid stenotype expression
+STENOTYPE = SPECIALS | CONTAINERS | LITERALS | SHORTHANDS
+
+#: any valid type expression
+TYPE << (STENOTYPE | TYPING)
+TYPE_exclude_UNION << (CONTAINERS | LITERALS | SHORTHANDS | ANY | OPTIONAL | TYPING)
+
+
+def parse(steno_string: str) -> ste.Steno:
```

Comment:
the `parse` and `unparse` functions look a little out of place here. I also don't like the whole `backend` structure that I imposed any more, it might make sense to delete that folder and move `elements.py` and `grammar.py` one level higher, rename `parser.py` to `parsing.py` and put these functions in there.

----

Regarding the functions, I am not sure I understand each of their purpose too well. I would have expected the following functionality:

 - map from stenotype string representation to the intermediate type form
 - map from standard typing string representation to the intermediate type form (not sure if this is really needed)
 - map from the intermediate form to stenotype string representation (I know this is just calling str on them, but I think it would be good to have a separate call interface for it)
 - map from the intermediate form to standard typing string representation

Is all of this already present, then I just need to read better or more documentation is needed.

Reply:
I agree that the current structure is difficult to follow. I suggest to keep the module structure as is for now, but clearly separate the parse/unparse actions.

``stenotype.backend.grammar.parse`` Converts *both* steno string -> IR and typing string -> IR, since typing strings are a subset of steno strings.  That one probably just needs documenting better. I do not see much point in moving the ``parse`` function since it is the "public API" of the  grammar.

The IR -> steno via ``str`` seems very cute/clever now. Having a separate, *explicit* call interface indeed seems clearer. We need a separate IR -> typing anyways.

-----------

The the IR is defined with ``singledispatch`` in mind, initially to support IR -> typing. That one would work for IR -> steno as well, of course. So I would recommend that I add a basic ``steno(:Steno) -> str`` now to replace ``str(:Steno)``.